### ISSUE TITLE
HVM AppVM sets time as UTC at boot, ignoring time zone offset

### ISSUE THREAD

**Issue by icequbes1** [QubesOS/qubes-issues#6280]
**Qubes OS version**
- R4.1
- Xen 4.14.0-8
- libvirt 6.6.0-2
- 5.4.80|5.9.12|5.18.16 dom0-provided kernel for AppVM

Not observed on R4.0, Xen 4.8.5-27, 5.8.16 kernel-latest, but minimal testing performed.

**Affected component(s) or functionality**
HVM AppVM date/time prior to qubes-sync-time service executing

**Brief summary**
HVM AppVM sets time as UTC at boot, though it appears the time is a local time being interpreted as UTC.

This means if local time is 8 hours behind UTC, the AppVM thinks local time is UTC-16 hours.

Assuming no ClockVM has been set in the dom0 Global Settings, this leaves an HVM hours ahead of or behind dom0 time based on the user's time zone. If time zone is behind UTC, this causes issues with "future files" existing.

This issue is the result of chasing down why an HVM did not have any journal logs after boot. systemd-journald at boot reported the journal file was corrupt - the assumption is the TemplateVM's journal was seen as a "future" file and deemed corrupt.

Therefore, an AppVM booted soon after its TemplateVM with localtime in the western hemisphere might observe this issue.

**How Reproducible**
As long as AppVM is set to HVM without any further customization. Issues may be more obvious if localtime is a negative offset from UTC.

**To Reproduce**

1. Set ClockVM to None in dom0 Qubes Global Settings
2. Boot an AppVM with virtualization mode set to HVM
3. Run `timedatectl` to observe time within the AppVM

**Expected behavior**
Correct date and local time displayed which (more or less) matches dom0 time.

**Actual behavior**
Date and time are displayed as local, but the local time is twice as far from UTC than what it should be.

**Screenshots**

True time is 19:00 UTC, 11am PST (-0800), what is observed:

```
dom0$ timedatectl
Local time: Sat 2020-12-12 11:00:00 PST
Universal time: Sat 2020-12-12 19:00:00 UTC

hvm$ timedatectl
Local time: Sat 2020-12-12 03:00:00 PST
Universal time: Sat 2020-12-12 11:00:00 UTC

# (observed in console logs at boot, minutes before)
localhost kernel: rtc_cmos 00:02: setting system clock to 2020-12-12T10:37:38 UTC

```

**Additional context**

ClockVM is purposely disabled. This means after boot, qubes-sync-time.service will run, but `qvm-sync-clock` will return "Invalid date received, aborting!" since no ClockVM is configured.

While this means the AppVM time may get skewed as it runs, my expectation is that an AppVM should boot with proper time provided by dom0's initial configuration of the VM.

#5616 pointed me to `/usr/share/qubes/templates/libvirt/xen.xml` and I see there is a `{% block clock %}` section with logic specific to an HVM, my HVM libvirt xml appears to ultimately get configured with:

```
<clock offset="variable" adjustment="0" basis="localtime" />
```

On R4.0, where I did not observe this issue, the same `<clock ... />` configuration is seen.

I have not set any `timezone` feature for the template/appvm, nor have I seen it documented anywhere, though [this qubes-doc](https://www.qubes-os.org/doc/standalone-and-hvm/#cloning-hvms) shows a `timezone` pref when grepping for `localtime` in qubes-doc.

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**

- #5616

**Comment by iamahuman** [comment]
Related: #1437, #1184, #5616

**Comment by DemiMarie** [comment]
Culprit appears to be this part of the template you mentioned:

```jinja
    {% block clock %}
        {% if vm.virt_mode == 'hvm' %}
            {% set timezone = vm.features.check_with_template('timezone', 'localtime').lower() %}
            {% if timezone == 'localtime' %}
                <clock offset="variable" adjustment="0" basis="localtime" />
            {% elif timezone.isdigit() %}
                <clock offset="variable" adjustment="{{ timezone }}" basis="utc" />
            {% else %}
                <clock offset="variable" adjustment="0" basis="utc" />
            {% endif %}
        {% else %}
            <clock offset='utc' adjustment='reset'>
                <timer name="tsc" mode="native"/>
            </clock>
        {% endif %}
    {% endblock %}
```

That is, qubesd checks if a timezone value has been set, and defaults to localtime if none has been. This had no affect in R4.0, because the old libvirt ignored that setting.  R4.1 honors it, however, so the qube’s emulated hardware clock is set the current _local_ time.  This results in Windows qubes behaving correctly in R4.1

However, Linux assumes that the hardware clock is set to UTC!  Therefore, what the qube _thinks_ is UTC time is actually _local_ time.  The qube knows what its timezone is, so it applies the timezone correction _again_, resulting in the qube’s local time differing from UTC by exactly twice the timezone correction.  This is especially bad because `sys-net` (the default ClockVM) is _always_ an HVM, so the resulting incorrect time is propagated to dom0, causing all sorts of badness.

The fix is simple: for non-Windows qubes, the hardware clock should **always** be set to UTC, irrespective of what the qube’s timezone is.  This can be checked by using the `os` feature, which will be set to `Windows` on Windows templates.  This will also require setting the `qubes-timezone` QubesDB entry according to the timezone, but that is a separate issue.

**Comment by DemiMarie** [comment]
Marking as P: critical because on systems using `systemd-resolved` for DNS resolution in `sys-net` (default in recent Fedora unless Qubes OS disables it), this causes DNSSEC validation to fail, preventing `chronyd` (and most likely `systemd-timesyncd` as well) from finding the NTP servers and resetting the clock.

**Comment by iamahuman** [comment]
> The fix is simple: for non-Windows qubes, the hardware clock should **always** be set to UTC, irrespective of what the qube’s timezone is. This can be checked by using the `os` feature, which will be set to `Windows` on Windows templates.

I suppose we can take this further and simply return UTC time regardless of which guest OS is in use. IMHO a better approach to the Windows local time problem would be instructing Windows read real-time clock as UTC, at the _same_ moment QWT advertises the OS to dom0. See also: #5616.

> This will also require setting the `qubes-timezone` QubesDB entry according to the timezone, but that is a separate issue.

**Comment by DemiMarie** [comment]
> I suppose we can take this further and simply return UTC time regardless of which guest OS is in use. IMHO a better approach to the Windows local time problem would be instructing Windows to read real-time clock as UTC, as QWT is responsible for advertising the OS to dom0 anyway. See also: #5616.

That is a good point.  The only remaining problem seems to be that the Windows NTP client breaks, but that isn’t an issue in Qubes OS as the NTP client runs in the ClockVM.

**Comment by iamahuman** [comment]
> > I suppose we can take this further and simply return UTC time regardless of which guest OS is in use. IMHO a better approach to the Windows local time problem would be instructing Windows to read real-time clock as UTC, as QWT is responsible for advertising the OS to dom0 anyway. See also: #5616.
> 
> That is a good point. The only remaining problem seems to be that the Windows NTP client breaks, but that isn’t an issue in Qubes OS as the NTP client runs in the ClockVM.

Also note that QWT currently leaves the timekeeping function in XenAgent enabled, which means it is oftentimes what ends up being the one setting the system clock anyway.

(XenAgent also has a problem with suspend/resume clock jumping, which RTC doesn't seem to share; this ought to be a separate issue, though)

**Comment by DemiMarie** [comment]
Fixed by QubesOS/qubes-core-admin#417.
